### ISSUE TITLE
Suggestions for rolling & series operations

### ISSUE THREAD

**Issue by walter-cavinaw** [databricks/koalas#1386]
first I wanted to say that I am impressed by what you are building and how rapidly you are iterating. I am an avid user of dask which is functionally very similar to koalas+pyspark combo.

I thought I would point out some of the things that would hold me back from switching over to using koalas. Maybe these are not actually problems or they can be circumvented, if so please let me know if I am mistaken.

I use a lot of rolling functions across partitions that are too large to be held in memory (on my machine), and also cumulative functions. Currently it seems koalas has to bring all data to a single partition to perform these operations. By default spark isn't able to handle rolling operations across partitions either unless you redefine the partitions yourself (again, please correct me if I am wrong.)

Dask does handle this out-of-the-box by increasing the size of each partition with the edge of its neighbor partition (see this link: https://docs.dask.org/en/latest/array-overlap.html).

The best solution would be for spark to do this natively, but the next best thing could be for koalas to use its knowledge of when a rolling operation is used to extend partitions so that the overlap.

Thoughts?

**Comment by HyukjinKwon** [comment]
> spark isn't able to handle rolling operations across partitions either unless you redefine the partitions yourself (again, please correct me if I am wrong.)

Yes, what you said is true. Koalas is currently implementing by sending all partitions to single note and calculating, which can easily end up with OOM (as documented). So, at this moment, it's preferred to perform rolling operations with groupby.

it is a good point to support rolling across multiple partitions.

Reading from Dask' doc:

> This costs an excess copy and the communication of many small chunks, but allows localized functions to evaluate in an embarrassingly parallel manner.

Seems they simply make a copy and send to another.

I am not sure, though, how we will perform the windoing in parallel because cumulative results depends on its previous results.

**Comment by walter-cavinaw** [comment]
For the cumulative functions, I don't think its necessary to be able to do it in parallel. It would be great just to be able to do it. Dask is also a task scheduler so I think the way it works is that it just perform the operation on partitions one after the other. Obviously you don't take advantage of parallelism, but at least I can perform cumulative functions across a dataset too big for memory.

Keep up the good work, I am excited to see the progress on koalas.
